### PR TITLE
Add a "deterministic end-to-end test" runner

### PR DESCRIPTION
**Pull request by TyOverby**
Part of fixing #5912 

Review of this PR happened over at https://github.com/dotnet/roslyn/pull/6020 where I was accidentally trying to merge into master instead of stabilization.

### CODE REVIEW

**Comment by TyOverby** [comment]
@jaredpar You had some comments on the [old version of this PR](https://github.com/dotnet/roslyn/pull/6020).  When you sign off on this, I'll merge.

**Comment by TyOverby** [comment]
@dotnet/roslyn-compiler 

Jared was reviewing this, but he is now on baby-duty, so could one more person do a quick passthrough?

File: src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        };
+
+        public class TempExecutionEnvironment {
+            public readonly TempDirectory tempDirectory;
+            public readonly string compilerDirectory;
+
+            public readonly string csharpCompilerClientExecutable;
+            public readonly string basicCompilerClientExecutable;
+            public readonly string compilerServerExecutable;
+            public readonly string buildTaskDll;
```
Comment (agocke) [inline src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs]: These fields should start with a capital letter if they're public.

File: src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            s_buildTaskDllSrc,
+            ResolveAssemblyPath("System.Reflection.Metadata.dll"),
+            ResolveAssemblyPath("Microsoft.CodeAnalysis.CSharp.dll"),
+            ResolveAssemblyPath("Microsoft.CodeAnalysis.VisualBasic.dll"),
+            Path.Combine(s_clientExecutableBasePath, CompilerServerExeName + ".config"),
+            Path.Combine(s_clientExecutableBasePath, "csc.rsp"),
+            Path.Combine(s_clientExecutableBasePath, "vbc.rsp")
+        };
+
+        public class TempExecutionEnvironment {
```
Comment (agocke) [inline src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs]: I think this should be `Disposable` and override `Dispose` to dispose the `tempDirectory`.

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        /// 
+        /// The default flags are "/shared /deterministic+ /nologo /t:library".
+        /// </summary>
+        /// <param name="source"> The source code for the program that will be compiled </param>
+        /// <param name="additionalFlags"> A string containing any additional compiler flags </param>
+        /// <returns> An array of bytes that were read from the compiled DLL</returns>
+        private byte[] CompileAndGetBytes(string source, string additionalFlags)
+        {
+            // Make a new execution environment for each compile so that each compile goes in its own directory.
+            var executionEnvironment = new ExecutingTestUtils.TempExecutionEnvironment();
```
Comment (agocke) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: You'll probably have to `using` this if you make the environment disposable.

File: src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs (hunk 1 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs]: Done.

File: src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs (hunk 2 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs]: Since tempDirectory is a disposable anyway, won't the behavior be unchanged?

Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs]: `TempDirectory` doesn't implement `Disposable`, but it has a private field `TempRoot` that _does_.  

I don't think there's a way to directly dispose the tempDirectory.

Comment (agocke) [inline src/Compilers/Core/VBCSCompilerTests/ExecutingTestUtils.cs]: Then I think `tempRoot` will need to be stored in a field as well and disposed.

**Comment by TyOverby** [comment]
@agocke: I've lost hours to the disposability issue.  I vote we leave it as-is for now and fix in the future.  This PR needs to land sooner rather than later because right now we have no full determinism tests.

**Comment by agocke** [comment]
@TyOverby OK, can you file a bug to rework this later? In this case I also vote against changing the VBCSCompiler tests. Just copy the code that's necessary to run your tests.

**Comment by TyOverby** [comment]
@agocke I don't think that's a good idea.  About 80% of ExecutingTestUtils.cs would just be duplicated in both files.

**Comment by TyOverby** [comment]
@agocke, Please review again.  Thanks for making me do it this way, it's definitely cleaner :+1:

**Comment by agocke** [comment]
:+1:

**Pull request merged**